staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go

```diff
@@ -327,14 +335,19 @@ func patchResource(
 			if err := applyPatchToObject(codec, defaulter, currentObjMap, originalPatchMap, versionedObjToUpdate, versionedObj); err != nil {
 				return nil, err
 			}
-			// Convert the object back to unversioned.
-			gvk := kind.GroupKind().WithVersion(runtime.APIVersionInternal)
-			objToUpdate, err := unsafeConvertor.ConvertToVersion(versionedObjToUpdate, gvk.GroupVersion())
```

Comment:
To understand the issue: was this the line which failed for CRDs?

I wonder whether we shouldn't fix the CR converter implementation insteads of fixing the symptoms here that CRs do not exist as internal version.

Also I wonder whether we shouldn't know the internal GVK here (which would match the external GVK for CRDs) instead of hardcoding APIVersionInternal.

@deads2k I guess you have seen this kind of issues before while implementing the CR strategy.

Reply:
> To understand the issue: was this the line which failed for CRDs?

yes.

Reply:
> I wonder whether we shouldn't fix the CR converter implementation insteads of fixing the symptoms here that CRs do not exist as internal version.

I found that CR converter only set the nested value in json. 
https://github.com/kubernetes/kubernetes/blob/7f9f847ce90d4a301a87d2f52cba1b9aa4859cf2/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go#L512

Reply:
> @deads2k I guess you have seen this kind of issues before while implementing the CR strategy.

I don't recall it.  Seems like it's worth expanding the apiextensions-apiserver's integration tests to cover patch.

Reply:
Done.
Added an integration test.
@sttts @deads2k

Reply:
Before merging this, I would like to see the alternative: add a artificial internal type to CRs by setting the api version in a given Unstructured to empty. Would this make all these changes unnecessary?

Rational: we started with CRDs to get rid of all the hacky exceptions in the core code for ThirdPartyResources. This kind of change goes the opposite direction again. I would pretty much prefer to invest in some hacks in apiextensions-apiserver over changing core code to cope with the CRD use-case.

Reply:
> Before merging this, I would like to see the alternative: add a artificial internal type to CRs by setting the api version in a given Unstructured to empty. Would this make all these changes unnecessary?

I did some research ahout this.
And I found a question [here](https://github.com/kubernetes/kubernetes/blob/4ff8cb60fa586ffb23d98bec5cb1ec892c39b2c9/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L322)

We can encode internal type for resources in the core code.
But we can't encode internal type for CRD. We don't know it's version.

Reply:
@sttts would it make more sense to do `unsafeConvertor.ConvertToVersion(versionedObjToUpdate, currentObject.GetObjectKind().GroupVersionKind())`? That's really what we want to end up with, whether currentObject was internal to start or not.

Reply:
I like that more than checking for versioned or not.

Reply:
(would need to check if `currentObject.GetObjectKind().GroupVersionKind()` returns the right thing for internal struct objects... it might return an empty `schema.GroupVersionKind`)

Reply:
It compiles only with `currentObject.GetObjectKind().GroupVersionKind().GroupVersion()`, since it needs to implement the `GroupVersioner` interface.

But fails:

```
--- FAIL: TestPatchResourceNumberConversion (0.00s)
	rest_test.go:221: Starting test TestPatchResourceNumberConversion
	rest_test.go:257: Working with patchType application/merge-patch+json
	rest_test.go:299: TestPatchResourceNumberConversion: unexpected error: v1.Pod is not suitable for converting to ""
	rest_test.go:257: Working with patchType application/strategic-merge-patch+json
	rest_test.go:299: TestPatchResourceNumberConversion: unexpected error: v1.Pod is not suitable for converting to ""
```

Reply:
right. I think TypeMeta just reports whatever Kind and APIVersion fields the struct contains, which for decoded internal types, is empty.

Could maybe do something like this:
```
// Convert to the current object's GVK
currentObjectGVK := currentObject.GetObjectKind().GroupVersionKind().GroupVersion()
if currentObjectGVK.Empty() {
  currentObjectGVK = kind.GroupKind().WithVersion(runtime.APIVersionInternal)
}
... = unsafeConvertor.ConvertToVersion(versionedObjToUpdate, currentObjectGVK)
```

Reply:
Yes, this works with `currentObjectGV = kind.GroupKind().WithVersion(runtime.APIVersionInternal).GroupVersion()`. 

Maybe we should call it `currentObjectGV` instead of `currentObjectGVK` since it's `GroupVersion`, not `GroupVersionKind`.

Reply:
yeah, mistyped